android/src/main/java/com/airbnb/android/react/maps/AirMapModule.java

```diff
@@ -0,0 +1,126 @@
+        final Integer height = options.hasKey("height") ? (int)(displayMetrics.density * options.getDouble("height")) : 0;
+        final String result = options.hasKey("result") ? options.getString("result") : "file";
+
+        // Add UI-block so we can get a valid reference to the map-view
+        UIManagerModule uiManager = context.getNativeModule(UIManagerModule.class);
+        uiManager.addUIBlock(new UIBlock() {
+            public void execute (NativeViewHierarchyManager nvhm) {
+                AirMapView view = (AirMapView) nvhm.resolveView(tag);
+                if (view == null) {
+                    promise.reject("AirMapView not found");
+                    return;
+                }
+                if (view.map == null) {
+                    promise.reject("AirMapView.map is not valid");
+                    return;
+                }
+                view.map.snapshot(new GoogleMap.SnapshotReadyCallback() {
+                    public void onSnapshotReady(Bitmap snapshot) {
+
+                        // Convert image to requested width/height if neccesary
+                        if (snapshot != null && width != 0 && height != 0 && (width != snapshot.getWidth() || height != snapshot.getHeight())) {
+                            snapshot = Bitmap.createScaledBitmap(snapshot, width, height, true);
+                        }
+                        if (snapshot == null) {
+                            promise.reject("Failed to generate bitmap");
+                            return;
+                        }
+
+                        // Save the snapshot to disk
+                        OutputStream outputStream = null;
+                        try {
+                            if (result.equals(SNAPSHOT_RESULT_FILE)) {
+                                File tempFile = File.createTempFile("AirMapSnapshot", "." + format, context.getCacheDir());
+                                outputStream = new FileOutputStream(tempFile);
+                                snapshot.compress(compressFormat, (int)(100.0 * quality), outputStream);
+                                outputStream.close();
+                                outputStream = null;
+                                String uri = Uri.fromFile(tempFile).toString();
+                                promise.resolve(uri);
+                            }
+                            else if (result.equals(SNAPSHOT_RESULT_BASE64)) {
+                                outputStream = new ByteArrayOutputStream();
+                                snapshot.compress(compressFormat, (int)(100.0 * quality), outputStream);
+                                outputStream.close();
+                                outputStream = null;
+                                byte[] bytes = ((ByteArrayOutputStream) outputStream).toByteArray();
+                                String data = Base64.encodeToString(bytes, Base64.NO_WRAP);
+                                promise.resolve(data);
```

Comment:
it would be nice if you could break down this logic into smaller methods. maybe `takeBase64Snapshot()` and `takeFileSnapshot()`?